ribbon-loadbalancer/src/main/java/com/netflix/loadbalancer/RandomRule.java

```diff
@@ -48,17 +48,18 @@ public Server choose(ILoadBalancer lb, Object key) {
             List<Server> upList = lb.getReachableServers();
             List<Server> allList = lb.getAllServers();
 
+            int upCount = upList.size();
             int serverCount = allList.size();
-            if (serverCount == 0) {
+            if (serverCount == 0 || upCount == 0) {
                 /*
-                 * No servers. End regardless of pass, because subsequent passes
+                 * No servers or No servers available. End regardless of pass, because subsequent passes
                  * only get more restrictive.
                  */
                 return null;
             }
 
             int index = chooseRandomInt(serverCount);
-            server = upList.get(index);
+            server = allList.get(index);

```

Comment:
Server should not be selected on upList ?
```
int index = chooseRandomInt(upCount);
server = upList.get(index);
```

Reply:
Yes, server should be selected on upList, but the index should not be chosen from serverCount, it may be greater than upList.size()